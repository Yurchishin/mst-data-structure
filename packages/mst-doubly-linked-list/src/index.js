export { default as DoublyLinkedList } from './DoublyLinkedList'
export { default as DoublyLinkedListNode } from './DoublyLinkedListNode'