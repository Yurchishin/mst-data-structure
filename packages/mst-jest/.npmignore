__test__
src
.babelrc
.eslintignore
.eslintrc
coverage
.gitignore
.npmignore
jest.config.js
rollup.config.js