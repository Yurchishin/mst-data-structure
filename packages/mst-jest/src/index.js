import * as mobxStateTree from 'mobx-state-tree';
import {onPatch, onSnapshot, getSnapshot } from 'mobx-state-tree';

export {
    protect,
    unprotect,
    isProtected,
} from 'mobx-state-tree';

export const createMSTSnapshot = (model, ...options) => {
    expect(getSnapshot(model)).toMatchSnapshot(...options);
};

export const getEnvMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getEnv').mockReturnValue(value);
};

export const getRootMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getRoot').mockReturnValue(value);
};

export const mockMiddleware = (call, next) => next(call);

export const onPatchesAggregation = (model, ...options) => {
    const patches = [];

    const disposer = onPatch(model, patch => {
        patches.push(patch);
    });

    return () => {
        expect(patches).toMatchSnapshot(...options);
        disposer();
    };
};

export const onSnapshotAggregation = (model, ...options) => {
    const snapshots = [];

    const disposer = onSnapshot(model, patch => {
        snapshots.push(patch);
    });

    return () => {
        expect(snapshots).toMatchSnapshot(...options);
        disposer();
    };
};
