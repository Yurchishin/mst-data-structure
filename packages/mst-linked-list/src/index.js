export { default as LinkedList } from './LinkedList'
export { default as LinkedListNode } from './LinkedListNode'