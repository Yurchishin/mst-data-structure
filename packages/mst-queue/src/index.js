export { default as Queue } from './Queue'
export { default as QueueNode } from './QueueNode'