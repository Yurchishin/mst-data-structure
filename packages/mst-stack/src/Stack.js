import { types, isReferenceType } from 'mobx-state-tree';
import LinkedList from '@mst/mst-linked-list';

const Stack = (name = '', Type, options = {}) => {
    if (isReferenceType(Type)) {
        throw new Error('"mst-stack" does not support reference types')
    }

    const StackModel = types
        .model(`${name}Stack`, {
            linkedList: LinkedList(`${name}Stack`, Type, options),
        })
        .views(self => ({
            get values() {
                return self.linkedList.values
            },
            get entries() {
                return self.linkedList.entries
            },
            get array() {
                return self.linkedList.array
            },
            get isEmpty() {
                return self.linkedList.isEmpty
            },
            toString(callback) {
                return self.linkedList.toString(callback)
            },
            get peek() {
                return !self.isEmpty ? self.linkedList.head : null
            },
            get peekPair() {
                return self.peek ? [self.peek.id, self.peek.value] : null
            },
            get peekValue() {
                return self.peekPair ? self.peekPair[1] : null
            },
            get peekId() {
                return self.peekPair ? self.peekPair[0] : null
            },
        }))
        .actions(self =>  {
            const push = (id, value) => {
                self.linkedList.prepend(id, value);
                return self
            }

            const pop = () => {
                self.linkedList.deleteHead();
                return self
            }

            return {
                push,
                pop,
            }
        })

    return types.optional(StackModel, {})
}

export default Stack;