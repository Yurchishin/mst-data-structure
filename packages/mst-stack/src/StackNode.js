import {isReferenceType, types} from "mobx-state-tree";

export const maybeNullReference = (subType, options = {}) => types.maybeNull(types.reference(subType, options))

const StackNode = (name, Type) => {
    if (isReferenceType(Type)) {
        throw new Error('"mst-stack" does not support reference types')
    }

    const NodeType = types
        .model(`${name}StackNode`, {
            id: types.identifier,
            value: Type,
            next: maybeNullReference(types.late(() => NodeType)),
        })
        .views(self => ({
            toString(callback) {
                return callback ? callback(self.value) : `${self.value}`;
            },
        }))
        .actions(self => {
            const setNext = nextNode => {
                self.next = nextNode;
            };

            return {
                setNext,
            };
        });

    return NodeType
}

export default StackNode