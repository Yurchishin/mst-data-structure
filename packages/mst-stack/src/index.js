export { default as Stack } from './Stack'
export { default as StackNode } from './StackNode'