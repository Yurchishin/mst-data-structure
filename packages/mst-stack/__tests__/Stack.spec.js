import { types } from 'mobx-state-tree';
import { onSnapshotAggregation, createMSTSnapshot } from '@mst-ds/mst-jest';
import { Stack, StackNode } from '../src/index';

describe('Stack', () => {
    describe('model', () => {
        it('create(int)', () => {
            const IntStackNode = StackNode('Int', types.integer)
            const list = Stack('Int', IntStackNode).create();

            createMSTSnapshot(list)
        });
        it('create in model(int)', () => {
            const IntStackNode = StackNode('Int', types.integer)

            const rootStore = types.model('RootStore', {
                numbers: Stack('Int', IntStackNode),
            }).create()

            createMSTSnapshot(rootStore)
        });
        it('create in model(Todo)', () => {
            const Todo = types.model('Todo', {
                name: types.string,
                active: types.boolean,
            })

            const TodoStackNode = StackNode('Todo', Todo)

            const rootStore = types.model('RootStore', {
                todos: Stack('Todo', TodoStackNode),
            }).create()

            rootStore.todos
                .push('1', {
                    name: 'name1',
                    active: false,
                })
                .push('2', {
                    name: 'name2',
                    active: true,
                })
                .push('3', {
                    name: 'name2',
                    active: false,
                })
                .pop()

            createMSTSnapshot(rootStore)
        });
    })
    describe('actions', () => {
        it('push', () => {
            const IntStackNode = StackNode('Int', types.integer)
            const list = Stack('Int', IntStackNode).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.isEmpty).toBe(true);

            list.push('3', 3);
            expect(list.peekNode.id).toBe('3');

            list.push('2', 2)
            expect(list.peekNode.id).toBe('2');

            list.push('1', 1)
            expect(list.peekNode.id).toBe('1');

            disposer()
        });
        it('pop', () => {
            const IntStackNode = StackNode('Int', types.integer)
            const list = Stack('Int', IntStackNode).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.isEmpty).toBe(true);

            list.push('3', 3)
                .push('2', 2)
                .push('1', 1)
                .pop()

            expect(list.peekNode.id).toBe('2');

            disposer()
        });
    });
});
