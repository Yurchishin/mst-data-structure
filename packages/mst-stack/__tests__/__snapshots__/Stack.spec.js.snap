// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack actions pop 1`] = `
Array [
  Object {
    "head": "3",
    "nodes": Object {
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
]
`;

exports[`Stack actions push 1`] = `
Array [
  Object {
    "head": "3",
    "nodes": Object {
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
  },
]
`;

exports[`Stack model create in model(Todo) 1`] = `
Object {
  "todos": Object {
    "head": "2",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": Object {
          "active": false,
          "name": "name1",
        },
      },
      "2": Object {
        "id": "2",
        "next": "1",
        "value": Object {
          "active": true,
          "name": "name2",
        },
      },
    },
  },
}
`;

exports[`Stack model create in model(int) 1`] = `
Object {
  "numbers": Object {
    "head": null,
    "nodes": Object {},
  },
}
`;

exports[`Stack model create(int) 1`] = `
Object {
  "head": null,
  "nodes": Object {},
}
`;
